omniledger/service/viewchange.go

```diff
@@ -0,0 +1,371 @@
+package service
+
+import (
+	"bytes"
+	"errors"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/dedis/cothority"
+	cosiprotocol "github.com/dedis/cothority/ftcosi/protocol"
+	"github.com/dedis/cothority/omniledger/darc"
+	"github.com/dedis/cothority/omniledger/viewchange"
+	"github.com/dedis/cothority/skipchain"
+	"github.com/dedis/kyber/sign/schnorr"
+	"github.com/dedis/onet"
+	"github.com/dedis/onet/log"
+	"github.com/dedis/onet/network"
+	"github.com/dedis/protobuf"
+)
+
+type viewChangeManager struct {
+	sync.Mutex
+	controllers map[string]*viewchange.Controller
+}
+
+func newViewChangeManager() viewChangeManager {
+	return viewChangeManager{
+		controllers: make(map[string]*viewchange.Controller),
+	}
+}
+
+func (m *viewChangeManager) add(sendVC viewchange.SendInitReqFunc,
+	sendNV viewchange.SendNewViewReqFunc, isLeader viewchange.IsLeaderFunc, k string) {
+	m.Lock()
+	defer m.Unlock()
+	c := viewchange.NewController(sendVC, sendNV, isLeader)
+	m.controllers[k] = &c
+}
+
+func (m *viewChangeManager) start(myID network.ServerIdentityID, scID skipchain.SkipBlockID, initialDuration time.Duration, f int, k string) {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[k]
+	go c.Start(myID, scID, initialDuration, f)
+}
+
+func (m *viewChangeManager) addReq(req viewchange.InitReq) {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[string(req.View.Gen)]
+	c.AddReq(req)
+}
+
+func (m *viewChangeManager) addAnomaly(req viewchange.InitReq) {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[string(req.View.Gen)]
+	c.AddAnomaly(req)
+}
+
+func (m *viewChangeManager) done(view viewchange.View) {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[string(view.Gen)]
+	c.Done(view)
+	log.Lvl3("view-change done for " + view.String())
+}
+
+func (m *viewChangeManager) waiting(k string) bool {
+	m.Lock()
+	defer m.Unlock()
+	c := m.controllers[k]
+	return c.Waiting()
+}
+
+func (m *viewChangeManager) closeAll() {
+	m.Lock()
+	defer m.Unlock()
+	for _, c := range m.controllers {
+		c.Stop()
+	}
+	m.controllers = make(map[string]*viewchange.Controller)
+}
+
+// sendViewChangeReq is called when the node detects that a view change is
+// needed. It uses SendRaw to send the message to all other nodes. This
+// function should only be used as a callback in ViewChangeLog.
+func (s *Service) sendViewChangeReq(view viewchange.View) error {
+	log.Lvl2(s.ServerIdentity(), "sending view-change request for view:", view)
+	latest, err := s.db().GetLatestByID(view.ID)
+	if err != nil {
+		return err
+	}
+	req := viewchange.InitReq{
+		SignerID: s.ServerIdentity().ID,
+		View:     view,
+	}
+	if err := req.Sign(s.getPrivateKey()); err != nil {
+		return err
+	}
+	for _, sid := range latest.Roster.List {
+		if sid.Equal(s.ServerIdentity()) {
+			continue
+		}
+		if err := s.SendRaw(sid, &req); err != nil {
```

Comment:
Good question, I don't know if `SendRaw` can block (most likely it can) or whether there's a way to set a timeout if it blocks for too long (I don't see one at the moment). We could put `SendRaw` in go-routines, which will make `sendViewChangeReq` non-blocking. But we will have no way to ask `SendRaw` to stop and we'll be left with lingering go-routines. I think there needs to be a version of `SendRaw` that supports a timeout.

Reply:
Added `there needs to be a version of SendRaw that supports a timeout` to 468, will do it real soon now.